File: rust/crates/test/src/ser.rs
use crate::schema;

#[test]
fn serialize_empty() {
  let m: schema::Test = schema::Empty {}.into();
  m.serialize();
}

#[test]
fn serialize_u8() {
  let m: schema::Test = schema::U8 { value: 69 }.into();
  let ser = m.serialize();
  assert!(ser[1] == 69);
  assert!(ser.len() == 2)
}

#[test]
fn serialize_u16() {
  let m: schema::Test = schema::U16 { value: 0x1234 }.into();
  let ser = m.serialize();
  assert!(ser[1] == 0x12);
  assert!(ser[2] == 0x34);
  assert!(ser.len() == 3)
}

#[test]
fn serialize_u32() {
  let m: schema::Test = schema::U32 { value: 0x12345678 }.into();
  let ser = m.serialize();
  assert!(ser[1] == 0x12);
  assert!(ser[2] == 0x34);
  assert!(ser[3] == 0x56);
  assert!(ser[4] == 0x78);
  assert!(ser.len() == 5)
}

#[test]
fn serialize_u64() {
  let m: schema::Test = schema::U64 {
    value: 0x123456789ABCDEF0,
  }
  .into();
  let ser = m.serialize();
  assert!(ser[1] == 0x12);
  assert!(ser[2] == 0x34);
  assert!(ser[3] == 0x56);
  assert!(ser[4] == 0x78);
  assert!(ser[5] == 0x9a);
  assert!(ser[6] == 0xbc);
  assert!(ser[7] == 0xde);
  assert!(ser[8] == 0xf0);
  assert!(ser.len() == 9)
}

#[test]
fn serialize_i8() {
  let m: schema::Test = schema::I8 { value: 69 }.into();
  let ser = m.serialize();
  assert!(ser[1] == 69);
  assert!(ser.len() == 2)
}

#[test]
fn serialize_i16() {
  let m: schema::Test = schema::I16 { value: 0x1234 }.into();
  let ser = m.serialize();
  assert!(ser[1] == 0x12);
  assert!(ser[2] == 0x34);
  assert!(ser.len() == 3)
}

#[test]
fn serialize_i32() {
  let m: schema::Test = schema::I32 { value: 0x12345678 }.into();
  let ser = m.serialize();
  assert!(ser[1] == 0x12);
  assert!(ser[2] == 0x34);
  assert!(ser[3] == 0x56);
  assert!(ser[4] == 0x78);
  assert!(ser.len() == 5)
}

#[test]
fn serialize_i64() {
  let m: schema::Test = schema::I64 {
    value: 0x123456789ABCDEF0,
  }
  .into();
  let ser = m.serialize();
  assert!(ser[1] == 0x12);
  assert!(ser[2] == 0x34);
  assert!(ser[3] == 0x56);
  assert!(ser[4] == 0x78);
  assert!(ser[5] == 0x9a);
  assert!(ser[6] == 0xbc);
  assert!(ser[7] == 0xde);
  assert!(ser[8] == 0xf0);
  assert!(ser.len() == 9)
}

#[test]
fn serialize_i8_negative() {
  let m: schema::Test = schema::I8 { value: -69 }.into();
  let ser = m.serialize();
  assert!(ser[1] == 187);
  assert!(ser.len() == 2)
}

#[test]
fn serialize_i16_negative() {
  let m: schema::Test = schema::I16 { value: -0x1234 }.into();
  let ser = m.serialize();
  assert!(ser[1] == 0xED);
  assert!(ser[2] == 0xCC);
  assert!(ser.len() == 3)
}

#[test]
fn serialize_i32_negative() {
  let m: schema::Test = schema::I32 { value: -0x12345678 }.into();
  let ser = m.serialize();
  assert!(ser[1] == 0xED);
  assert!(ser[2] == 0xCB);
  assert!(ser[3] == 0xA9);
  assert!(ser[4] == 0x88);
  assert!(ser.len() == 5)
}

#[test]
fn serialize_i64_negative() {
  let m: schema::Test = schema::I64 {
    value: -0x123456789ABCDEF0,
  }
  .into();
  let ser = m.serialize();
  assert!(ser[1] == 0xED);
  assert!(ser[2] == 0xCB);
  assert!(ser[3] == 0xA9);
  assert!(ser[4] == 0x87);
  assert!(ser[5] == 0x65);
  assert!(ser[6] == 0x43);
  assert!(ser[7] == 0x21);
  assert!(ser[8] == 0x10);
  assert!(ser.len() == 9)
}

#[test]
fn serialize_string() {
  let m: schema::Test = schema::Str {
    value: "hello!".into(),
  }
  .into();
  let ser = m.serialize();
  assert!(ser[1] == 0x68);
  assert!(ser[7] == 0x0);
  assert!(ser.len() == 8)
}

#[test]
fn serialize_f32() {
  let m: schema::Test = schema::F32 {
    value: std::f32::consts::PI,
  }
  .into();
  let ser = m.serialize();
  assert!(ser[1] == 0x40);
  assert!(ser[2] == 0x49);
  assert!(ser[3] == 0x0f);
  assert!(ser[4] == 0xdb);
  assert!(ser.len() == 5)
}

#[test]
fn serialize_f64() {
  let m: schema::Test = schema::F64 {
    value: std::f64::consts::PI,
  }
  .into();
  let ser = m.serialize();
  assert!(ser[1] == 0x40);
  assert!(ser[2] == 0x09);
  assert!(ser[3] == 0x21);
  assert!(ser[4] == 0xFB);
  assert!(ser[5] == 0x54);
  assert!(ser[6] == 0x44);
  assert!(ser[7] == 0x2D);
  assert!(ser[8] == 0x18);
  assert!(ser.len() == 9)
}

#[test]
fn serialize_zero_termination() {
  let m: schema::Test = schema::ZeroTermination {
    a: "hello!".into(),
    b: 0x69,
  }
  .into();
  let ser = m.serialize();
  assert!(ser.len() == 9);
  assert!(ser[7] == 0);
}

File: lua/tests/ser.lua
local Test = require("tests.gen.test")

return {
	testSerializeEmpty = function()
		local msg = Test.Empty.new():_wrap()
		msg:_serialize()
	end,

	testSerializeU8 = function()
		local msg = Test.U8.new(69):_wrap()
		local ser = msg:_serialize()
		Lu.assertEquals(ser:sub(2, 2):byte(), 69)
		Lu.assertEquals(#ser, 2)
	end,

	testSerializeBool = function()
		local msg = Test.Bool.new(true):_wrap()
		local ser = msg:_serialize()
		Lu.assertEquals(ser:sub(2, 2):byte(), 1)
		Lu.assertEquals(#ser, 2)
	end,

	testSerializeU16 = function()
		local msg = Test.U16.new(0x1234):_wrap()
		local ser = msg:_serialize()
		Lu.assertEquals(ser:sub(2, 2):byte(), 0x34)
		Lu.assertEquals(ser:sub(3, 3):byte(), 0x12)
		Lu.assertEquals(#ser, 3)
	end,

	testSerializeU32 = function()
		local msg = Test.U32.new(0x12345678):_wrap()
		local ser = msg:_serialize()
		Lu.assertEquals(ser:sub(2, 2):byte(), 0x78)
		Lu.assertEquals(ser:sub(3, 3):byte(), 0x56)
		Lu.assertEquals(ser:sub(4, 4):byte(), 0x34)
		Lu.assertEquals(ser:sub(5, 5):byte(), 0x12)
		Lu.assertEquals(#ser, 5)
	end,

	testSerializeI8 = function()
		local msg = Test.I8.new(69):_wrap()
		local ser = msg:_serialize()
		Lu.assertEquals(ser:sub(2, 2):byte(), 69)
		Lu.assertEquals(#ser, 2)
	end,

	testSerializeI16 = function()
		local msg = Test.I16.new(0x1234):_wrap()
		local ser = msg:_serialize()
		Lu.assertEquals(ser:sub(2, 2):byte(), 0x34)
		Lu.assertEquals(ser:sub(3, 3):byte(), 0x12)
		Lu.assertEquals(#ser, 3)
	end,

	testSerializeI32 = function()
		local msg = Test.I32.new(0x12345678):_wrap()
		local ser = msg:_serialize()
		Lu.assertEquals(ser:sub(2, 2):byte(), 0x78)
		Lu.assertEquals(ser:sub(3, 3):byte(), 0x56)
		Lu.assertEquals(ser:sub(4, 4):byte(), 0x34)
		Lu.assertEquals(ser:sub(5, 5):byte(), 0x12)
		Lu.assertEquals(#ser, 5)
	end,

	testSerializeI8Negative = function()
		local msg = Test.I8.new(-69):_wrap()
		local ser = msg:_serialize()
		Lu.assertEquals(ser:sub(2, 2):byte(), 187)
		Lu.assertEquals(#ser, 2)
	end,

	testSerializeI16Negative = function()
		local msg = Test.I16.new(-0x1234):_wrap()
		local ser = msg:_serialize()
		Lu.assertEquals(ser:sub(2, 2):byte(), 0xcc)
		Lu.assertEquals(ser:sub(3, 3):byte(), 0xed)
		Lu.assertEquals(#ser, 3)
	end,

	testSerializeI32Negative = function()
		local msg = Test.I32.new(-0x12345678):_wrap()
		local ser = msg:_serialize()
		Lu.assertEquals(ser:sub(2, 2):byte(), 0x88)
		Lu.assertEquals(ser:sub(3, 3):byte(), 0xa9)
		Lu.assertEquals(ser:sub(4, 4):byte(), 0xcb)
		Lu.assertEquals(ser:sub(5, 5):byte(), 0xed)
		Lu.assertEquals(#ser, 5)
	end,

	testSerializeString = function()
		local msg = Test.Str.new("hello!"):_wrap()
		local ser = msg:_serialize()
		Lu.assertEquals(ser:sub(2, 2):byte(), 0x68)
		Lu.assertEquals(ser:sub(8, 8):byte(), 0x00)
		Lu.assertEquals(#ser, 8)
	end,

	testSerializeF32 = function()
		local msg = Test.F32.new(math.pi):_wrap()
		local ser = msg:_serialize()
		Lu.assertEquals(ser:sub(2, 2):byte(), 0xda)
		Lu.assertEquals(ser:sub(3, 3):byte(), 0x0f)
		Lu.assertEquals(ser:sub(4, 4):byte(), 0x49)
		Lu.assertEquals(ser:sub(5, 5):byte(), 0x40)
		Lu.assertEquals(#ser, 5)
	end,

	testSerializeF64 = function()
		local msg = Test.F64.new(math.pi):_wrap()
		local ser = msg:_serialize()
		Lu.assertEquals(ser:sub(2, 2):byte(), 0x18)
		Lu.assertEquals(ser:sub(3, 3):byte(), 0x2d)
		Lu.assertEquals(ser:sub(4, 4):byte(), 0x44)
		Lu.assertEquals(ser:sub(5, 5):byte(), 0x54)
		Lu.assertEquals(ser:sub(6, 6):byte(), 0xfb)
		Lu.assertEquals(ser:sub(7, 7):byte(), 0x21)
		Lu.assertEquals(ser:sub(8, 8):byte(), 0x09)
		Lu.assertEquals(ser:sub(9, 9):byte(), 0x40)
		Lu.assertEquals(#ser, 9)
	end,

	testSerializeZeroTermination = function()
		local msg = Test.ZeroTermination.new("hello!", 69):_wrap()
		local ser = msg:_serialize()
		Lu.assertEquals(ser:sub(8, 8):byte(), 0x00)
		Lu.assertEquals(#ser, 9)
	end,
}
